### PR TITLE
feat: code action

### PR DESCRIPTION
**Pull request by d-jeffery**
## Description

* Enabled code actions
* Ported over some actions from `syntax-transformer`

## References

// TODO: Testing

## Review Checklist
- [ ] I have clicked on ["allow edits by maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [ ] I have added documentation for new/changed functionality in this PR or in a PR to [openfga.dev](https://github.com/openfga/openfga.dev) [Provide a link to any relevant PRs in the references section above]
- [x] The correct base branch is being used, if not `main`
- [ ] I have added tests to validate that the change in functionality is working as expected

### CODE REVIEW

File: server/src/server.common.ts [hunk 1]
```diff
@@ -152,6 +185,51 @@ export function startServer(connection: _Connection) {
...
+		}
+		
+		// Determine indent
+		let indent = "\t";
+		if (insertSpaces) {
+			indent = " ";
+			for (let i = 1; i < tabSize; i++) {
+				indent += " ";
+			}
+		}
```
Comment (d-jeffery) [inline server/src/server.common.ts:219]: Just realized in testing that indents are either 2 spaces or a tab. I will remove this logic.

I will default to 2 spaces, or a single tab based on settings.

File: server/src/code-action.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+	docUri: DocumentUri,
+	diagnostic: Diagnostic
+}
+
+export function getMissingDefinitionFix(params: CodeActionParams, indent: string): CodeAction {
+
+	const textDocEdit: TextDocumentEdit = {
+		textDocument: OptionalVersionedTextDocumentIdentifier.create(params.docUri, null),
+		edits: [
+			TextEdit.insert({ line: params.diagnostic.range.start.line, character: 0 },
```
[the remark this answers is not in the document]
Comment (rhamzeh) [reply server/src/code-action.ts:21]: Should we be inserting on the line after instead of before?

[the remark this answers is not in the document]
Comment (d-jeffery) [reply server/src/code-action.ts:21]: I had that originally, but you cant insert on a line that doesnt exist, so you cant do `line + 1`.

I think I can change it to capture the current line, and rewrite via replace with the correct new lines.

File: server/src/server.common.ts [hunk 3]
```diff
@@ -83,6 +91,19 @@ export function startServer(connection: _Connection) {
...
+	async function updateIndentationConfig() {
+		const config = await connection.workspace.getConfiguration();
+		tabSize = config.editor.tabSize;
+		insertSpaces = config.editor.insertSpaces;
+
+		if (config["[openfga]"] && config["[openfga]"]["editor.tabSize"] !== undefined) {
+			tabSize = config["[openfga]"]["editor.tabSize"];
+		}
+		if (config["[openfga]"] && config["[openfga]"]["editor.insertSpaces"] !== undefined) {
+			insertSpaces = config["[openfga]"]["editor.insertSpaces"];
```
[the remark this answers is not in the document]
Comment (d-jeffery) [reply server/src/server.common.ts:98]: So i dont think we can affect changes to configuration from within the language server.

I have simplified the indentation to: `const indent = insertSpaces ? "  " : "\t";`

We might be able to prompt the user to change their settings from the extension tho.

**Review by rhamzeh — APPROVED** [review]

**Pull request merged**